include/engine/api/nearest_api.hpp

```diff
@@ -38,8 +38,50 @@ class NearestAPI final : public BaseAPI
                        phantom_nodes.front().end(),
                        waypoints.values.begin(),
                        [this](const PhantomNodeWithDistance &phantom_with_distance) {
-                           auto waypoint = MakeWaypoint(phantom_with_distance.phantom_node);
+                           auto &phantom_node = phantom_with_distance.phantom_node;
+                           auto waypoint = MakeWaypoint(phantom_node);
                            waypoint.values["distance"] = phantom_with_distance.distance;
+
+                           util::json::Array nodes;
+
+                           std::uint64_t from_node = static_cast<std::uint64_t>(SPECIAL_OSM_NODEID);
+                           std::uint64_t to_node = static_cast<std::uint64_t>(SPECIAL_OSM_NODEID);
+
+                           auto segment_id = phantom_node.forward_segment_id.id;
+                           if (segment_id != SPECIAL_SEGMENTID)
+                           {
+                               const auto geometry_id = facade.GetGeometryIndex(segment_id).id;
+                               std::vector<NodeID> geometry =
+                                   facade.GetUncompressedForwardGeometry(geometry_id);
+
+                               auto osm_node_id = facade.GetOSMNodeIDOfNode(
+                                   geometry[phantom_node.fwd_segment_position]);
+                               to_node = static_cast<std::uint64_t>(osm_node_id);
+                           }
+
+                           segment_id = phantom_node.reverse_segment_id.id;
+                           if (segment_id != SPECIAL_SEGMENTID)
+                           {
+                               const auto geometry_id = facade.GetGeometryIndex(segment_id).id;
+                               std::vector<NodeID> geometry =
+                                   facade.GetUncompressedForwardGeometry(geometry_id);
+                               auto osm_node_id = facade.GetOSMNodeIDOfNode(
+                                   geometry[phantom_node.fwd_segment_position + 1]);
+                               from_node = static_cast<std::uint64_t>(osm_node_id);
+                           }
+
+                           if (!phantom_node.forward_segment_id.enabled)
+                           {
+                               nodes.values.push_back(to_node);
+                               nodes.values.push_back(from_node);
+                           }
+                           else
+                           {
+                               nodes.values.push_back(from_node);
+                               nodes.values.push_back(to_node);
+                           }
+                           waypoint.values["nodes"] = std::move(nodes);
```

Comment:
Could you assert we never return the special osm id sentinel set above here please

Reply:
Yeah we should not expose the osrm internal special node id to the outside. We should probably just ail with an error code in this edge case.

Reply:
Hrm, would be interesting to see how this could happen. From my understanding there is a guarantee that at least one of `forward` and `backward` is enabled. The code in `nearest.cpp` should already bail out with `NoSegement` if we didn't found candidates.

Reply:
Most of edge case are one way case. In that case `reverse_segment_id.enabled` is false. 

I wonder if it would not be better to use `forward_segment_id` only and get the previous node like this
```
auto osm_node_id = facade.GetOSMNodeIDOfNode(geometry[phantom_node.fwd_segment_position - 1]);
```
As @danpat commented in #2581 , the edge case would be when `fwd_segment_position` is zero.  I guess this would require to find segments linked to `forward_segment_id` to access to their geometry. 
Do you think this would be correct?

Reply:
Any feedback on my comment?